### DISCUSSION TITLE
Wordpress CMS example preview mode

### DISCUSSION THREAD

**Discussion by janhoogeveen** [vercel/next.js#13982] [Help]
First up, massive thanks to @gregrickaby and @lfades for the Wordpress CMS example. I've been playing around with it, and it's a rock solid foundation for new projects.

One thing's been nagging me though... for preview mode to work, you need to:

1. Make a GraphQL API call
2. Save the refreshToken in your Next.js config
3. Restart your server

While these 3 steps doesn't sound like a lot of work, I don't think my clients will love me for proposing this workflow.

Bear in mind I haven't touched Wordpress in 8 years and my PHP knowledge is terrible, so I don't really know where to start. Still, I've been wondering if there's an easier solution for granting preview mode temporarily. 

Here's some open questions from my end to which I _really_ don't have an answer, but maybe this helps spark a conversation.

1. Would it be possible to use a [nonce](https://codex.wordpress.org/WordPress_Nonces) to grant a 15 minute read-only token for unpublished posts/drafts? This blogpost seems to indicate it should be possible. https://github.com/justinwhall/wordpress-gatsby-preview-starter
2. Would it be possible to open up certain endpoints with a static password token? E.g. can we use `NEXT_EXAMPLE_CMS_WORDPRESS_PREVIEW_SECRET` as the only token required to preview posts?

**Comment by lfades** [comment]
I created the WordPress example without any knowledge of WordPress before that (I learned it on the go). I agree that our preview setup is not ideal, it doesn't feel as smooth as in the other CMS examples, so if you can improve it, feel free to create a PR 🙏 

@gregrickaby Do you have any thoughts on this? 🕵️

**Comment by janus-reith** [comment]
@janhoogeveen If there is no better option, you could use https://github.com/WP-API/Basic-Auth

Although originally intended for the REST API, it also works fine with WP GraphQL.
You could use some generic API-User for Wordpress, pass the credentials to your nextjs env and thereby gain some flexibility how to handle things from there. You would only need some auth check in your nextjs api, and if that succedes it would just use the wordpress auth data. For said auth check inside nextjs, you could also define a static secret to pass, not the most secure solution but a really simple one.

Regarding the Nonce: Wordpress creates a Nonce when entering preview mode, and Ideally I'd like to just reuse that, together with some custom WP Hook to rewrite the preview url to the nextjs storefront.
But the Nonce only works together with additional data stored in a cookie and that's where I didn't get further yet.
Possible options I see:
A. Make the Wordpress Cookie Nonce available cross site and read It in your API endpoint.
B. Hook into the WP Preview URL Generation and create some custom Nonce that could work for API request.(To add more security, maybe include some secret only known to nextjs.) - That one is a bit blurry to me yet.

I'm still looking for a more clean way using the oauth Flow, but in the meantime the Basic auth approach at least allows for some flexibility.

**Comment by gregrickaby** [comment]
@janhoogeveen @janus-reith @lfades I like the idea of leveraging Basic Auth. I'd like to take this question back to my team (whom are much smarter than me) and see if they have any thoughts.

**Comment by janhoogeveen** [comment]
Another approach might be:

1. Wordpress preview links to /api/preview
2. NextJS preview redirects to a login page if you're not logged in through the front-end yet
3. Use JWT token to log in on front-end, redirect to a preview page
4. Use token to pick up preview.

This way you don't need to rely on a JWT token in your env vars, and the user won't have to use GraphiQl or something to set a token.

I'm gonna look into this solution right now.

**Reply by janhoogeveen** [reply]
Alright, a couple of hours later and I've got a proof of concept working:

In `/api/preview.js`, I perform a login action and place the refreshToken in the previewData cookie.

```diff
	import { loginUser } from '../../lib/api';
	import { getPreviewPost } from '../../lib/api';
	
	export default async function preview(req, res) {
      ...

+	  const user = await loginUser('user', 'password');
	
	  // Fetch WordPress to check if the provided `id` or `slug` exists
	  const post = await getPreviewPost(
	    id || slug,
	    id ? 'DATABASE_ID' : 'SLUG',
+       user.refreshToken
	  );
	
	  res.setPreviewData({
	    post: {
	      id: post.databaseId,
	      slug: post.slug,
	      status: post.status,
	    },
+	    refreshToken: user.refreshToken,
	  });

	  ...
	}

```

In `api.js` I changed `fetchApi` to support the refreshToken.

```diff
+	import Cookies from 'js-cookie';
	
+	export async function fetchAPI(query, { variables, refreshToken } = {}) {
	  const headers = { 'Content-Type': 'application/json' };
	
+	  const cookie = Cookies.get('auth');
	
+	  if (refreshToken) {
+	    headers['Authorization'] = `Bearer ${refreshToken}`;
+	  }
	
+	  if (cookie) {
+	    const auth = JSON.parse(cookie);
+	    headers['Authorization'] = `Bearer ${auth.refreshToken}`;
+	  }
      ...
	}
```

I modified the `getPreviewPost` and `getPostAndMorePosts` methods to accept a refreshToken as well, which get passed to `fetchApi`.

As a proof of concept, I'm really happy with this. Obviously the fact that the login is hardcoded isn't great, but there must be a way to improve on that.

**Next steps?**
- [x] Add a login page
- [x] Redirect users to the login page if they're trying to open a preview route while not being logged in
- [x] Store the logged in users name and refreshToken in a cookie and in memory. A cookie is great for client-side requests, but afaik getStaticProps and API routes can not access cookies. Is that correct @lfades?
- [ ] Use the cookie/memory in getStaticProps instead of passing the refreshToken around.

**Reply by lfades** [reply]
@janhoogeveen `getStaticProps` can only access one cookie, and that's for preview mode, while in preview mode `getStaticProps` will work like a serverless function, you can detect this and use `previewData` to save context. Other custom cookies won't work.

**Reply by janhoogeveen** [reply]
> @janhoogeveen `getStaticProps` can only access one cookie, and that's for preview mode, while in preview mode `getStaticProps` will work like a serverless function, you can detect this and use `previewData` to save context. Other custom cookies won't work.

Yep, thanks. Sounds like I understood correctly. That means I'll to pass the refreshToken through the preview cookie, but I don't think that's a problem.

**Reply by janus-reith** [reply]
@janhoogeveen I didn‘t get yet how the term refreshToken is used here. It seems like this is used as an access token? I would expect the refresh token to enable me to get a new access token, instead of having just that refresh token to authorize.

**Reply by MoOx** [reply]
Same here, I don't really understand how the example is supposed to work...

**Comment by kpoelhekke** [comment]
I've been able to get this working by passing the refreshToken as a query param via the Wordpress preview link. This way you don't need to work with login credentials in Next.

This assumes you have the `JWT Auth – WordPress JSON Web Token Authentication` plugin installed. I don't use GraphQL in my setup but that shouldn't make a difference.

```php
<?php
namespace JWTAuth;

add_filter('preview_post_link', function ($preview_link, $post) {
    $auth = new Auth();

    $user = wp_get_current_user();

    $token = $auth->generate_token($user);

    return add_query_arg([
        'post_type' => $post->post_type,
        'post_id' => $post->ID,
        'token' => $token
    ], home_url() . '/api/preview');
}, 10, 2);
```

**Reply by hmtri1011** [reply]
Hi @kpoelhekke, I came from google and followed your code, but there has an issue on my end.

I did install the plugin (version 1.2.4) but when I defined namespace and constructed a new Auth() instance, the error appeared. Can you help me? Sorry, I'm not familiar with WP

**Reply by hmtri1011** [reply]
This is what I did to make it work

```
add_filter('preview_post_link', function ($preview_link, $post) {
  $user = wp_get_current_user();

  // Generate token with jwt-auth
  $url = home_url() . '/wp-json/jwt-auth/v1/token/get';

  $results = wp_remote_post($url, array(
    'headers' => array(
      'content-type' => 'application/json'
    ),
    'body' => json_encode(array(
      'user_id' => $user->ID
    ))
  ));
  $results = json_decode($results['body']);

  $token = $results->token;
  $expire = $results->expire;

  return add_query_arg([
      'post_id' => $post->ID,
      'secret' => $token
  ], home_url() . '/api/preview');
}, 10, 2);
```

**Reply by pascalroget** [reply]
am I wrong in feeling like passing a JWT token as a GET parameter is a potential security issue?

**Comment by MoOx** [comment]
I tried to follow the next.js wordpress example but I can't get anything working when it comes to auth. The example is indeed weird because refreshToken are supposed to only give you a new authToken... 
Beside this I think the readme is missing some informations because on my end, it seems my query made for preview doesn't work as the no post come in the response (but if I use a post that is published, preview works - I understand here that the auth is not effective so wordpress doesn't give me back post that are not published).

Does anyone have something solid to share? poke @janhoogeveen

**Reply by pascalroget** [reply]
A bit late to reply, but are you passing the Authorization header in your query?
If using Apollo, it is done like this:

```javascript
const response = await client.query({
query: GET_POST,
variables: {
id: isDraft ? postPreview.id : slug,
idType: isDraft ? "DATABASE_ID" : "SLUG",
},
context: {
headers: {
Authorization: Bearer ${process.env.WORDPRESS_AUTH_REFRESH_TOKEN},
},
},
});

````
